clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -6203,7 +6217,102 @@ void SYCLIntegrationHeader::emit(raw_ostream &O) {
   O << "} // namespace detail\n";
   O << "} // namespace _V1\n";
   O << "} // namespace sycl\n";
-  O << "\n";
+
+  unsigned ShimCounter = 1;
+  int FreeFunctionCount = 0;
+  for (const KernelDesc &K : KernelDescs) {
+    if (!isFreeFunction(S, K.SyclKernel))
+      continue;
+
+    ++FreeFunctionCount;
+    // Generate forward declaration for free function.
+    O << "\n// Definition of " << K.Name << " as a free function kernel\n";
+    if (K.SyclKernel->getLanguageLinkage() == CLanguageLinkage)
+      O << "extern \"C\" ";
+    std::string ParmList;
+    bool FirstParam = true;
+    for (ParmVarDecl *Param : K.SyclKernel->parameters()) {
+      if (FirstParam)
+        FirstParam = false;
+      else
+        ParmList += ", ";
+      ParmList += Param->getType().getCanonicalType().getAsString();
+    }
+    FunctionTemplateDecl *FTD = K.SyclKernel->getPrimaryTemplate();
+    if (FTD) {
+      Policy.SuppressDefinition = true;
+      FTD->print(O, Policy);
+      O << ";\n";
+    } else {
```

Comment:
I am confused here. Why is this `else `required? Why can't you use print for everything?
Also, should you be printing `K.SYCLKernel `and not the primary template?

Reply:
Using "print" by itself for non-templated functions produces weird results. For instance, the attributes applied to the function (add_ir_attributes_function values) are printed in a garbled manner with pointer values in hex for some fields. That's why there are two different paths. One for templated functions and the other for non-templated functions.
For templated functions the intent is to first generate a forward declaration of the template. Then, a cast expression of a specific instantiation of the templated function can be applied to a function address in the corresponding shim function.
Look at what's done for function ff_3 in the test free_function_int_header.cpp. For an "int" instantiation the integration header contains this:
```
template <typename T> void ff_3(T *ptr, T start, T end);
static constexpr auto __sycl_shim3() {
  return (void (*)(int *, int, int))ff_3<int>;
}
```
Other instantiations of the same function will lead to different shims with different casts.

Reply:
>>Using "print" by itself for non-templated functions produces weird results. For instance, the attributes applied to the function (add_ir_attributes_function values) are printed in a garbled manner with pointer values in hex for some fields.

So attributes are printed for non-templated functions but not printed for templates? This seems odd. Can you also please share what is printed here ? Did you debug to check if it was a printing policy mismatch/why print behaves differently?

Reply:
Using the same print function for non-templated functions produces these forward declarations : 
```
// Definition of _Z18__sycl_kernel_ff_2Piii as a free function kernel
[[__sycl_detail__::add_ir_attributes_function(0x556831a325f0, 0x556831a32650)]] __attribute__((sycl_device)) void ff_2(int *ptr, int start, int end);
static constexpr auto __sycl_shim1() {
  return (void (*)(int *, int, int))ff_2;
}

// Definition of _Z18__sycl_kernel_ff_2Piiii as a free function kernel
[[__sycl_detail__::add_ir_attributes_function(0x556831a34f80, 0x556831a34fe0)]] __attribute__((sycl_device)) void ff_2(int *ptr, int start, int end, int value);
static constexpr auto __sycl_shim2() {
  return (void (*)(int *, int, int, int))ff_2;
}
```

Reply:
I looked into setting some other Policy flags and found that `PolishForDeclaration `cleans up the declarations of non-templated functions. So that's what I've done now.